consensus/spos/consensusMessageValidator.go

```diff
@@ -0,0 +1,270 @@
+	}
+
+	if len(cnsMsg.PubKey) != core.PublicKeySizeInBytes {
+		return fmt.Errorf("%w : received public key from consensus topic has an invalid size: %d",
+			ErrInvalidPublicKeySize,
+			len(cnsMsg.PubKey))
+	}
+
+	if len(cnsMsg.Signature) != core.SignatureSizeInBytes {
+		return fmt.Errorf("%w : received signature from consensus topic has an invalid size: %d",
+			ErrInvalidSignatureSize,
+			len(cnsMsg.Signature))
+	}
+
+	isNodeInEligibleList := wrk.consensusState.IsNodeInEligibleList(string(cnsMsg.PubKey))
+	if !isNodeInEligibleList {
+		return fmt.Errorf("%w : received message from consensus topic has an invalid public key: %s",
+			ErrNodeIsNotInEligibleList,
+			logger.DisplayByteSlice(cnsMsg.PubKey))
+	}
+
+	if wrk.consensusState.RoundIndex+1 < cnsMsg.RoundIndex {
+		log.Trace("received message from consensus topic has a future round",
+			"msg type", wrk.consensusService.GetStringValue(msgType),
+			"from", cnsMsg.PubKey,
+			"header hash", cnsMsg.BlockHeaderHash,
+			"msg round", cnsMsg.RoundIndex,
+			"round", wrk.consensusState.RoundIndex,
+		)
+
+		return ErrMessageForFutureRound
+	}
+
+	if wrk.consensusState.RoundIndex > cnsMsg.RoundIndex {
+		log.Trace("received message from consensus topic has a past round",
+			"msg type", wrk.consensusService.GetStringValue(msgType),
+			"from", cnsMsg.PubKey,
+			"header hash", cnsMsg.BlockHeaderHash,
+			"msg round", cnsMsg.RoundIndex,
+			"round", wrk.consensusState.RoundIndex,
+		)
+
+		return ErrMessageForPastRound
+	}
+
+	err = wrk.checkSignature(cnsMsg)
+	if err != nil {
+		return fmt.Errorf("%w : verify signature for received message from consensus topic failed: %s",
+			ErrInvalidSignature,
+			err.Error())
+	}
+
+	return nil
+}
+
+func (wrk *Worker) checkConsensusMessageValidityForMessageType(cnsMsg *consensus.Message) error {
+	msgType := consensus.MessageType(cnsMsg.MsgType)
+
+	if wrk.consensusService.IsMessageWithBlockBodyAndHeader(msgType) {
+		return wrk.checkMessageWithBlockBodyAndHeaderValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithBlockBody(msgType) {
+		return wrk.checkMessageWithBlockBodyValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithBlockHeader(msgType) {
+		return wrk.checkMessageWithBlockHeaderValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithSignature(msgType) {
+		return wrk.checkMessageWithSignatureValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithFinalInfo(msgType) {
+		return wrk.checkMessageWithFinalInfoValidity(cnsMsg)
+	}
+
+	return fmt.Errorf("%w : received message type from consensus topic is invalid: %d",
+		ErrInvalidMessageType,
+		msgType)
+}
+
+func (wrk *Worker) checkMessageWithBlockBodyAndHeaderValidity(cnsMsg *consensus.Message) error {
+	isMessageInvalid := cnsMsg.SignatureShare != nil ||
+		cnsMsg.PubKeysBitmap != nil ||
+		cnsMsg.AggregateSignature != nil ||
+		cnsMsg.LeaderSignature != nil
+
+	if isMessageInvalid {
+		return fmt.Errorf("%w : received message from consensus topic is invalid: "+
```

Comment:
is this error message not too long ? maybe do a log trace before then send and simpler wrapped error.

Reply:
Done

Reply:
funny fact in GO - || has precedence over &&